core/src/main/java/org/jruby/RubyModule.java

```diff
@@ -1763,6 +1749,20 @@ public DynamicMethod searchMethodInner(String id) {
         return null;
     }
 
+    public CacheEntry searchMethodEntryInner(String id) {
+        int token = generation;
+        // This flattens some of the recursion that would be otherwise be necessary.
+        // Used to recurse up the class hierarchy which got messy with prepend.
+        for (RubyModule module = this; module != null; module = module.getSuperClass()) {
+            // Only recurs if module is an IncludedModuleWrapper.
+            // This way only the recursion needs to be handled differently on
+            // IncludedModuleWrapper.
+            DynamicMethod method = module.searchMethodCommon(id);
+            if (method != null) return method.isNull() ? null : cacheEntryFactory.newCacheEntry(id, method, module, token);
```

Comment:
the module.searchMethodCommon in RubyModule will return null, NullMethod, *Method, but in IncludedModuleWrapper.searchMethodCommon we look for Null method so it is only null, *Method.  I think we should have all versions behave as latter since we reduce number of cases we need to consider.

Reply:
Yeah I was just mimicking the existing logic throughout this stuff but it should be audited to see whether we still need all these representations of null.

Reply:
It looks like @tduehr introduced NullMethod when he did the first prepend impl for JRuby. @tduehr Was this just an attempt to start eliminating null from the searchMethod* logic? There's still only the one place that ever returns NullMethod so I'm wondering whether it's worth keeping.

Reply:
I needed to differentiate between a failed search and an empty result to break a recursion loop or something like that. I don't remember exactly what, but it shouldn't be too hard to swap it out for a null check and see if that breaks anything.